tests/integration_tests/functional/test_snapshot_advanced.py

```diff
@@ -0,0 +1,259 @@
+              NetIfaceConfig(host_ip="192.168.2.1",
+                             guest_ip="192.168.2.2",
+                             tap_name="tap2",
+                             dev_name="eth2"),
+              NetIfaceConfig(host_ip="192.168.3.1",
+                             guest_ip="192.168.3.2",
+                             tap_name="tap3",
+                             dev_name="eth3")]
+# Define 4 scratch drives.
+scratch_drives = ["vdb", "vdc", "vdd", "vde", "vdf"]
+
+
+@pytest.mark.skipif(
+    platform.machine() != "x86_64",
+    reason="Not supported yet."
+)
+def test_restore_old_snapshot_all_devices(bin_cloner_path):
+    """Test scenario: restore all previous version snapshots."""
+    # Microvm: 2vCPU 256MB RAM, baloon, 4 disks and 4 netdevices.
+    logger = logging.getLogger("snapshot_many_devices")
+
+    artifacts = ArtifactCollection(_test_images_s3_bucket())
+    # Fetch all firecracker binaries.
+    # With each binary create a snapshot and try to restore in current
+    # version.
+    firecracker_artifacts = artifacts.firecrackers()
+    for firecracker in firecracker_artifacts:
+        firecracker.download()
+        jailer = firecracker.jailer()
+        jailer.download()
+
+        logger.info("Creating snapshot with Firecracker: %s",
+                    firecracker.local_path())
+        logger.info("Using Jailer: %s", jailer.local_path())
+
+        target_version = firecracker.base_name()[1:]
+
+        # v0.23 does not support creating diff snapshots.
+        # v0.23 does not support balloon
+        diff_snapshots = "0.23" not in target_version
+
+        # Create a snapshot.
+        snapshot = create_snapshot_helper(bin_cloner_path,
+                                          logger,
+                                          drives=scratch_drives,
+                                          ifaces=net_ifaces,
+                                          fc_binary=firecracker.local_path(),
+                                          jailer_binary=jailer.local_path(),
+                                          diff_snapshots=diff_snapshots,
+                                          balloon=diff_snapshots)
+
+        # Resume microvm using current build of FC/Jailer.
+        builder = MicrovmBuilder(bin_cloner_path)
+        microvm, _ = builder.build_from_snapshot(snapshot,
+                                                 resume=True,
+                                                 enable_diff_snapshots=False)
+        validate_all_devices(logger, microvm, net_ifaces, scratch_drives,
+                             diff_snapshots)
+        logger.debug("========== Firecracker restore snapshot log ==========")
+        logger.debug(microvm.log_data)
+
+
+@pytest.mark.skipif(
+    platform.machine() != "x86_64",
+    reason="Not supported yet."
+)
+def test_restore_old_version_all_devices(bin_cloner_path):
+    """Test scenario: restore all previous version snapshots."""
```

Comment:
Create current version snapshot and restore in older/previous versions?

Reply:
Yes, it tries to restore all previous version snapshots in current version.

Reply:
I will update the description.

Reply:
When creating a snapshot, the Firecracker API allows to provide a target_version (the microVM version for which we want to create the snapshot). This is optional and defaults to the current version if not specified. Firecracker v0.24 supports v0.23.0 as target version for snapshotting. 

This test creates snapshots using the current Firecracker version, but targeting the old supported versions. Then, the snapshots are resumed using the binary artifacts.